lapack/unit_test/Test_Lapack_geqrf.hpp

```diff
@@ -0,0 +1,602 @@
+// CUDA+CUSOLVER, HIP+ROCSOLVER and HOST+LAPACK
+#if (defined(TEST_CUDA_LAPACK_CPP) && defined(KOKKOSKERNELS_ENABLE_TPL_CUSOLVER)) || \
+    (defined(TEST_HIP_LAPACK_CPP) && defined(KOKKOSKERNELS_ENABLE_TPL_ROCSOLVER)) || \
+    (defined(KOKKOSKERNELS_ENABLE_TPL_LAPACK) &&                                     \
+     (defined(TEST_OPENMP_LAPACK_CPP) || defined(TEST_SERIAL_LAPACK_CPP) || defined(TEST_THREADS_LAPACK_CPP)))
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Random.hpp>
+
+#include <KokkosBlas2_ger.hpp>
+#include <KokkosBlas3_gemm.hpp>
+#include <KokkosLapack_geqrf.hpp>
+#include <KokkosKernels_TestUtils.hpp>
+
+namespace Test {
+
+template <class ViewTypeA, class ViewTypeTau>
+void getQR(int const m, int const n, typename ViewTypeA::HostMirror const& h_A,
+           typename ViewTypeTau::HostMirror const& h_tau, typename ViewTypeA::HostMirror& h_Q,
+           typename ViewTypeA::HostMirror& h_R, typename ViewTypeA::HostMirror& h_QR) {
+  using ScalarA = typename ViewTypeA::value_type;
+
+  // ********************************************************************
+  // Populate h_R
+  // ********************************************************************
+  for (int i(0); i < m; ++i) {
+    for (int j(0); j < n; ++j) {
+      if (i <= j) {
+        h_R(i, j) = h_A(i, j);
+      } else {
+        h_R(i, j) = Kokkos::ArithTraits<ScalarA>::zero();
+      }
+    }
+  }
+
+  // ********************************************************************
+  // Instantiate the m x m identity matrix h_I
+  // ********************************************************************
+  ViewTypeA I("I", m, m);
+  typename ViewTypeA::HostMirror h_I = Kokkos::create_mirror_view(I);
+  Kokkos::deep_copy(h_I, Kokkos::ArithTraits<ScalarA>::zero());
+  for (int i(0); i < m; ++i) {
+    if constexpr (Kokkos::ArithTraits<ScalarA>::is_complex) {
+      h_I(i, i).real() = 1.;
+    } else {
+      h_I(i, i) = 1.;
+    }
+  }
+
+  // ********************************************************************
+  // Compute h_Q
+  // ********************************************************************
+  int minMN(std::min(m, n));
+  ViewTypeTau v("v", m);
+  typename ViewTypeTau::HostMirror h_v = Kokkos::create_mirror_view(v);
+
+  ViewTypeA Qk("Qk", m, m);
+  typename ViewTypeA::HostMirror h_Qk = Kokkos::create_mirror_view(Qk);
+
+  ViewTypeA auxM("auxM", m, m);
+  typename ViewTypeA::HostMirror h_auxM = Kokkos::create_mirror_view(auxM);
+
+  // Q = H(0) H(1) . . . H(min(M,N)-1), where for k=0,1,...,min(m,n)-1:
+  //   H(k) = I - Tau(k) * v * v**H, and
+  //   v is a vector of size m with:
+  //     v(0:k-1) = 0,
+  //     v(k)     = 1,
+  //     v(k+1:m-1) = A(k+1:m-1,k).
+  for (int k(0); k < minMN; ++k) {
+    Kokkos::deep_copy(h_v, Kokkos::ArithTraits<ScalarA>::zero());
+    h_v[k] = 1.;
+    for (int index(k + 1); index < m; ++index) {
+      h_v[index] = h_A(index, k);
+    }
+#if 0  // def HAVE_KOKKOSKERNELS_DEBUG
```

Comment:
Is it worthwhile to keep these `#if 0` blocks around? Should we switch these to something like `#if QEQRF_DEBUG` so that the programmer can turn these all back on easily?